packages/authentication/test/unit/provider.test.ts

```diff
@@ -0,0 +1,88 @@
+  let provider: Provider<AuthenticateFn>;
+  let strategy: MockStrategy;
+
+  const mockUser: UserProfile = {name: 'mock-user', locale: 'mock-locale'};
+
+  beforeEach(givenAuthenticationProvider);
+
+  describe('constructor()', () => {
+    it('instantiateClass injects authentication.strategy in the constructor', async () => {
+      const context: Context = new Context();
+      context.bind('authentication.strategy').to(strategy);
+      provider = await instantiateClass<Provider<AuthenticateFn>>(AuthenticationProvider, context);
+      const authenticationProvider: AuthenticationProvider = provider as AuthenticationProvider;
+      expect(authenticationProvider.getStrategy()).to.be.equal(strategy);
+    });
+  });
+
+  describe('value()', () => {
+    it('returns a function which authenticates a request and returns a user', async () => {
+      const authenticate: AuthenticateFn = await Promise.resolve(provider.value());
+      const request = <ParsedRequest> {};
+      const user = await authenticate(request);
+      expect(user).to.be.equal(mockUser);
+    });
+
+    describe('context.get(provider_key)', () => {
+      it('returns a function which authenticates a request and returns a user', async () => {
+        const context: Context = new Context();
+        context.bind('authentication.strategy').to(strategy);
+        context.bind('authentication.provider').toProvider(AuthenticationProvider);
+        const authenticate = await context.get('authentication.provider');
+        const request = <ParsedRequest> {};
+        const user = await authenticate(request);
+        expect(user).to.be.equal(mockUser);
+      });
+
+      it('throws an error if the injected passport strategy is not valid', async () => {
+        const context: Context = new Context();
+        context.bind('authentication.strategy').to({});
+        context.bind('authentication.provider').toProvider(AuthenticationProvider);
+        const authenticate = await context.get('authentication.provider');
+        const request = <ParsedRequest> {};
+        let error;
+        try {
+          await authenticate(request);
+        } catch (exception) {
+          error = exception;
+        }
+        expect(error.message).to.equal('invalid strategy parameter');
+      });
+
+      it('throws Unauthorized error when authentication fails', async () => {
+        const context: Context = new Context();
+        context.bind('authentication.strategy').to(strategy);
+        context.bind('authentication.provider').toProvider(AuthenticationProvider);
+        const authenticate = await context.get('authentication.provider');
+        const request = <ParsedRequest> {};
+        request.headers = {testState: 'fail'};
+        let error;
+        try {
+          const user: Object = await authenticate(request);
+        } catch (err) {
+          error = err;
+        }
+        expect(error).to.be.instanceof(HttpErrors.Unauthorized);
```

Comment:
I would rather not rely on inheritance here. The code handling HTTP errors does not care about the constructor of the error, it checks `statusCode` property. I think we should do the same here.

```ts
expect(error).to.have.property('statusCode', 401);
```

This way the test is decoupled from implementation details of the `authenticate` function and if we decide to change the way how `authenticate` reports errors in the future, then this test will keep working.